File: 2015/utils/expr.rb
def expr(eqn, vals)
  filled = eqn.map{|v| 
    case v 
    when 'AND'; '&'
    when 'OR'; '|'
    when 'LSHIFT'; '<<'
    when 'RSHIFT'; '>>'
    when 'NOT'; '~'
    else
      v.to_i.to_s == v ? v : vals[v]
    end
  }.join(' ')
  eval(filled) % 65536
end

File: 2015/utils/expr.js
const expr = (eqn, vals) => {
  const filled = eqn.map(v => {
    switch (v) {
      case 'AND': return '&'
      case 'OR': return '|'
      case 'LSHIFT': return '<<'
      case 'RSHIFT': return '>>'
      case 'NOT': return '~'
      default:
        return Number.isNumber(v) ? parseInt(v) : vals[v]
    }
  })
  return eval(filled.join(' ')).mod(65536)
}

export default expr